### PR TITLE
[AutoDiff] Add differentiation benchmarks

### PR DESCRIPTION
**Pull request by rxwei**
Revives #31108 by @marcrasi.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci benchmark

File: stdlib/public/runtime/AutoDiffSupport.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@ using namespace llvm;
 
 SWIFT_CC(swift)
 static void destroyLinearMapContext(SWIFT_CONTEXT HeapObject *obj) {
+  static_cast<AutoDiffLinearMapContext *>(obj)->~AutoDiffLinearMapContext();
```
Comment (marcrasi) [inline stdlib/public/runtime/AutoDiffSupport.cpp:22]: Is this related?

**Review by marcrasi — APPROVED** [review]

File: stdlib/public/runtime/AutoDiffSupport.cpp (hunk 1 above)
Comment (rxwei) [reply stdlib/public/runtime/AutoDiffSupport.cpp:22]: This is on top of #34900. The diff should disappear once it's merged.

**Review by atrick — APPROVED** [review]
lgtm

**Comment by rxwei** [comment]
@swift-ci benchmark

**Comment by rxwei** [comment]
@swift-ci benchmark

**Review by eeckstein — APPROVED** [review]
lgtm, though it seems that the benchmarks are crashing during run.

**Comment by rxwei** [comment]
Interesting... I'm unable to reproduce this on macOS.

**Comment by rxwei** [comment]
@swift-ci benchmark

**Comment by rxwei** [comment]
@swift-ci benchmark

**Comment by rxwei** [comment]
@edymtt any concerns with https://github.com/apple/swift/pull/34901/commits/3e86dc9396a8177eadeed7d2bc8ebdf6f0dc94ea?

**Comment by rxwei** [comment]
@swift-ci please test
